### DISCUSSION TITLE
ch7/ch7-15

### DISCUSSION THREAD

**Discussion by giscus** [gopl-zh/gopl-zh.github.com#166] [General]
# ch7/ch7-15

<The Go Programming Language>中文版

https://gopl-zh.github.io/ch7/ch7-15.html

**Comment by zhangbest5** [comment]
这是中文？想表达个啥？老太太裹脚布

**Reply by LiuDoo** [reply]
翻译确实一言难尽

**Reply by runningwa** [reply]
看的贼难受

**Comment by lazybearlee** [comment]
1. **新手Go程序员的常见做法**：新手程序员在设计新包时，往往会首先定义一套接口，然后创建一些具体类型来实现这些接口。这种做法可能导致很多接口，每个接口只有一个实现。

2. **避免不必要的抽象**：作者建议不要这样做，因为这种做法可能会产生不必要的抽象。接口的使用应该基于实际需要，而不是为了抽象而抽象。

3. **运行时损耗**：接口的使用会带来一定的运行时损耗，因为接口调用涉及到动态类型检查和间接调用。

4. **使用导出机制**：Go语言提供了导出机制（exporting），允许开发者控制类型的方法或结构体字段是否在包外可见。这意味着，如果只需要限制访问，不一定需要使用接口。

5. **接口的合理使用**：接口应该在有两个或两个以上的具体类型需要以相同的方式处理时才使用。这样可以避免过度设计，并保持接口的简洁性。

6. **接口作为解耦手段的例外**：如果由于依赖关系，一个具体类型不能和它实现的接口在同一个包中，那么接口可以作为这两个包之间的解耦手段。

7. **接口设计的抽象性**：由于接口通常需要从特定的实现细节中抽象出来，它们往往具有更少和更简单的方法。这样的接口更容易被新的类型实现。

8. **接口设计的指导原则**：一个好的接口设计应该遵循 "ask only for what you need"（只考虑你需要的东西）的原则，即接口应该只包含实现特定功能所需的最小方法集。

总结来说，这段文字强调了在Go语言中设计接口时应遵循的一些原则，包括避免不必要的抽象、合理使用接口、以及接口设计应简洁且只包含必要的方法。这些原则有助于创建更清晰、更易于维护的代码。

**Comment by Apahno** [comment]
这段文字主要想表达以下几点：

不要滥用接口：在设计包时，不要先创建一套接口，然后再定义具体类型。只有当有两个或以上的具体类型需要以相同的方式进行处理时，才需要定义接口。
接口应该是必要的抽象：接口应该是抽象的，不应该有太多的方法或字段。只有当两个或以上的类型实现一个接口时，才使用接口。
小接口更好：小接口更容易满足，尤其是当新的类型出现时。好的接口设计标准是“只考虑你需要的东西”。
Go语言支持多种编程风格：Go语言支持面向对象风格的编程，但也支持其他风格，如函数式编程。不是任何事物都需要被当做一个对象；独立的函数和未封装的数据类型也有它们自己的用处。

**Comment by Harlan6** [comment]
接口一个很重要的特性就是，他能解耦，屏蔽细节，使得使用更加简洁；所以对于只有一个实现，也是可以的，至少它清晰了交互协议
